### ISSUE TITLE
Icecrown Citadel

### ISSUE THREAD

**Issue by michalpolko** [mangosR2/scriptdev2#25]
It is no real issue, but since we have no r2 community forums... ;D

I am currently in progress of rewriting all ICC bosses that are implemented in BSWAI in mangosR2 (I write scripts in ScriptedAI) but maybe there is somebody interested in developing Icecrown Citadel, which is in fact the last and, well, the most important imho raid dungeon, so it would be very nice to have it completed as much as we can. ;)

What I have already done:
- Lord Marrowgar
- Lady Deathwhisper
- Deathbringer Saurfang
- Festergut
- Rotface
- Dreamwalker
- Blood Prince Council
- Blood Queen
- Sindragosa

TODO:
- Professor Putricide
- Lich King

I am not planning on scripting not-bosses (any events, minibosses, quests etc.). Maybe there is somebody who is interested in such stuff and would like to contribute? There are many events to do and some mobs need fixing spells / need some minor scripting, so everything would look nicely and much blizzlike.

I will push my scripts to mangosR2 when I will be sure they are finished (right now I have them tested on my repo/server).

So again, if somebody is interested in taking part of improving Icecrown Citadel - write here, so we will know who is working at what part of the instance, and maybe ask for help or anything, post links to commits etc. Let's work together! ;)

P.S. It is like fallenangel's work with Sunwell Plateau :) (good job btw)

**Comment by wolfiestyle** [comment]
I got a (nearly) full ICC implementation using a custom framework. Was wat i had on my old (now offline) server. It's like 1.5 years old and extensively tested. Not fully blizzlike and has some hacks, and not all scripts are vehicle-aware. At leas all bosses are doable (gunship not implemented by obvious reasons).

**Comment by ghost** [comment]
i agree were the elite and the leaders of public repo of WoW Emulation on the internet (world)  all our instances,quest ect ect should be there does anyone really have any idea how many people use our shit 8P

ive been following mich ICC work and was actually going to implent it myself after SWP is finished ( did have a reqeust for the chessboard event 8P) id rather mich do it 8P .. im down to help ive already been fixing minor sd2 and some major DB issues in ICC recently  quest event AI ect ect i down to help   but i will say  i plan on finish Swp or at least getting the last and first event playable for i cant focus really on anything else it there finished 

but as one of R2Sd2 main commiters i say push them if they work  we'll need to get the sql to fix up everyone Db ec t ect

**Comment by wolfiestyle** [comment]
just uploaded my old server's SD2 into this repo:
https://github.com/darkstalker/pwow_scriptdev2

most scripts written with the framework inside include/sc_utility.{h,cpp}
contains ICC, RS, TotC, Occulus, Violet Hold, and other stuff i don't remember now

**Comment by michalpolko** [comment]
Another library for scriptdev2 ;D First glance, looks like it has similiar boundries as bswai. Maybe some data will be helpful, thx for this.

**Comment by wolfiestyle** [comment]
class EventManager is something everyone should use imho... the "if (a < uiDiff) { ... } else a -= uiDiff" constructs are evil.
At least they took my idea of storing guid's in containers.

**Comment by rsa** [comment]
sended to me corrects to Valithria mobs:
http://pastebin.com/Cx3hEmd8
i not check this. may be needed.

**Comment by michalpolko** [comment]
I am not interested in bswai fixes, so I will not use it ;P

For me the current mangos and sd2 framework for scripting is enough, just need add a few methods for selecting targets in mangos (for filling targets map and in sd2 for single target spells) and that's all we need.

**Comment by rsa** [comment]
I'm not going to promote bswAI, it's bad enough the thing made ​​(as I wrote it, I still do not know C + +, just plain C). but the design mentioned DarkStalker, "stitching" timers in the code of the script - even more ugly and wrong path. can you come up new way, instead of using a blunt 5 years old approach ScriptedAI (or a year old bswAI?)

**Comment by ghost** [comment]
i dont think we need a new way  just some improvements to  ScriptedAI ( or some genius things like here GUIDS are staored  ( personally i like both  scriptedai and bsw) .. we dont need a new system  we need to start getting things done 8P

**Comment by michalpolko** [comment]
Using pseudo cool frameworks, which are just wrappers for the parent classes' methods doesn't improve coding skills...

**Comment by rsa** [comment]
write own mega-cool framework :) now tools/methods in SD2 simply ridiculous, and does not satisfy at least the minimum requirements...

**Comment by ghost** [comment]
8(   id have to learn a new system  i suck as it is now lol

**Comment by wolfiestyle** [comment]
maybe something with this interface:

m_events.add(Event::create().startsIn(1000).repeatsEvery(2500).onlyInPhase(2));

or some template driver stuff.. needs a good design. For me there are some clear components: timers, object summon/storage, instance handling, messaging (multi-script coordination)

**Comment by rsa** [comment]
may be get event syntax from TC, but store parameters in DB like BSW? not like hardcoded events like (2500). long time ago i start write trinity-mangos script connector, but not finish...

**Comment by ghost** [comment]
If were gonna do all that why don't we take bsw add what we like and use from scriptedai and improve from there

**Comment by michalpolko** [comment]
> m_events.add(Event::create().startsIn(1000).repeatsEvery(2500).onlyInPhase(2));

looks like doctrine query builder ;P

EventAI > all and that's all folks ^^

My opinion: the only thing that would be nice to implement in sd2 is handling timers in one place. if ()...else... is taking too much code and makes script less readable.
An event system would be nice. Passing timers to some TimersHandler in the AI constructor (also possibility of modifying timers separately during script execution) that would handle expiring the timer and execute given event handler - function.

**Comment by wolfiestyle** [comment]
EventAI is retarded, database is for storing _data_, not code. Trying to hack code into the relational model is just not right.

Try looking at the EventManager class in the repo i posted avobe, it's pretty complete. But that interface is based on TC's EventMap (original purpose was for porting TC scripts, but evolved later).

**Comment by michalpolko** [comment]
and creature_ai_scripts stores data, what code are You talking about? code is in mangos core. in DB You are storing data when and under what circumstances to perform which action and on whom. You would like writing all small and simple scripts for every mob in world in some c++ code instead of taking required data from DB and put it into script template?

If it is retarded, then seems like blizzard is retarded because wow is using such stuff afaik and as far as we can see in spell.dbc file - also for bosses (targeting and all effects are handled by core, not script, script is just "cast spell every now and then" and that's all).

P.S. retarded is using phrase "retarded"...

**Comment by ghost** [comment]
personally i like using a hybrid of ways to write scripts  ACID is one way  trash and things like that  or if its simple enough it should be done in ACID

**Comment by carlos93** [comment]
We have some icc bosses like Blood Queen working 75% - 80%, if you want to see https://github.com/Radicalmangos/devscriptdev2 (it is full spanish :()

**Comment by ghost** [comment]
8) ive actually been keeping an eye on both you and mich stuff lately  alot of good work going on there it just saddens me that where all divided work 8)

**Comment by carlos93** [comment]
If michalpolko push anything related with ICC, I pull it haha. His scripts are better but I try to learn :D.

**Comment by rsa** [comment]
as i see, this implementation has less work things, then abandoned by me a 1,5 years ago :)
PS don't use calls like this:
DoCastSpellIfCan(m_creature, m_bIs25Man ? SPELL_INFEST_25H : SPELL_INFEST_10H)
- seems as joke (obsolete and not worked as intended over a year).

**Comment by michalpolko** [comment]
use casts of normal spell id only for spells introduced after 3.2.0 patch, map difficulty based spell id is chosen in core when casting

**Comment by carlos93** [comment]
mmm ok, i don't know it :D. I have to rewrite a few scripts...

**Comment by Walkum** [comment]
Hi all, I'm the companion of carlos, I still learning in this world, very newbie :), michalpolko rlz for learn +1, your scripts are very great, gratz man ^^, rsa, your lich_king.cpp was improved by me, only the outro event and intro, is very blizzlike now, if I can send to R2 repo, tell me how I can, thanks ^^

**Comment by carlos93** [comment]
Is there anyone working in core support for the gunship battle?

**Comment by ghost** [comment]
ive actually been waiting on npc on transports commit was suppose to implented monthes ago but some of R2 devs have gone MIA 8) 

https://github.com/bwsrv/mangos/tree/transport  psz was working on it

**Comment by carlos93** [comment]
https://github.com/Radicalmangos/devscriptdev2/commit/750317b039f09f08b90bc180468bc7936092ddf5

This would be interesting. It has something wrong?

**Comment by michalpolko** [comment]
Some bunch of news and updates ;)

Currently I have made some further improvements in long ago scripted encounters so now they work even better.
Some still have a few things to test and fix.

Right now I am finishing Queen Lana'thel. Some spells are already implemented so the work is not so hard now ;)

So after I'm done with the queen only Professor Putricide and the Lich King will remain.
As I was thinking some time ago, I should be able to get some time to work with the LK around mid-late December, so he should be finished on New Year's Eve ;}

BUT about what I have changed my mind, is _when_ to push my ICC implementation to mangosR2 repository.
At first I wanted to finish it completely, so that I don't give You a half-product.

But now I am in need of help ;D I will push my work into separate branches so that You can test it and choose if You like it or not and use old BSWAI scripts instead. And I will do this when I'm done with Putricide, so that only Lick King will remain undone.

I hope that this will help me get some feedback, because right now on my test server there aren't too many players and they don't really want to help as much as I would like them to ;P So I will be asking all of You who uses r2 repository to test as much as You can and report any bugs found here, so we can have as perfect implementation of Icecrown Citadel as it is possible for us ;}
This will be around middle of December (I hope I will find time for it ;) ).

Until then, stay tuned ;)

**Comment by Walkum** [comment]
I can help you with feedback ;)

**Comment by carlos93** [comment]
We used your scripts in my server. Now we are working on improve stability and then we'll open raids. I can test your scripts and give you a feedback if I see something wrong or another thing ;).

**Comment by ghost** [comment]
I can lend a hand

**Comment by rsa** [comment]
> At first I wanted to finish it completely

@michalpolko, products like mangos code or SD2 scripts _nowhere_ must be finish completely :)

**Comment by michalpolko** [comment]
but I like getting things done ;)

**Comment by michalpolko** [comment]
OK, I finally decided to merge my ICC into r2.

I created branches for mangos and sd2. You can merge them, test and provide some feedback ;)

Putricide is not fully implemented yet. Abomination can't attack I think (faction problems or some vehicle flags), variables in heroic mode arent implemented, unbound plague on heroic can't be applied more than once on a given target (some aura restriction, needs research).
Lich King is not yet rewritten. I don't know when I will find will and time for this.

https://github.com/mangosR2/mangos/commit/6f0edcfa40e2268c5c32b81b329bd81614a596f6
https://github.com/mangosR2/scriptdev2/commit/fe0363151f47b72c7f2514891434978cafad4d18

**Comment by ghost** [comment]
TY mich

**Comment by carlos93** [comment]
Thx for the scripts :). I'll try to rewrite the LK (now that you haven't got enough time) based in your script style, and later you can improve if you want.

**Comment by rsa** [comment]
good news before holidays :) i think, this  must be immediately poured into the main branch, other tests will be carried out too few people...

**Comment by michalpolko** [comment]
Does anybody have working solution for Necrotic Plague spell?

**Comment by rsa** [comment]
see on this - https://gist.github.com/1593475
this very old solution, and not fully worked (cycled, as i remember), but partially works.
now need rework over spell_linked.

**Comment by michalpolko** [comment]
This is wrong.

Problem is setting stack amount which differs for different remove modes (reduce stack by 1 if dispelled, add 1 stack if any other remove mode).

Current auras system seems not to allow such mechanic with those 2 spells (main spell and second spell which stacks after jumps) because when I apply the "jump" aura and get caster - I can't obtain info about the aura from which plague jumped and dont know how many stacks to set :/

**Comment by rsa** [comment]
this spells cannot stack (from one caster) really. after any other remove method (except dispell) spell cause 50000+ damage without reducing.

**Comment by michalpolko** [comment]
?

When aura 70337 or 70338 is dispelled, its stack is reduced by 1 if stackamount > 1 and aura other than initial 70337

If 70338 jumps with other remove mode then it gains 1 stack (1 more than it was on the target before).

**Comment by michalpolko** [comment]
Another day of fighting with Necrotic Plague and still can't find the way to handle stacks depending on dispel type... Sigh...

**Comment by carlos93** [comment]
If you dispell (or another way) the debuff jumps? I finished that and it stacks (I need to see this better)

**Comment by michalpolko** [comment]
After dispel the stack must be reduced by 1, if remove mode is any other then the stack must be increased by 1.

**Comment by carlos93** [comment]
Yes, I know the mechanical :P

Actually I have "working"a beta for this spell but it needs a rewrite.

**Comment by michalpolko** [comment]
Could You show the code?

Also, last night I came up with an idea why my implementation didnt work, Ill try to fix this tonight ;]

**Comment by carlos93** [comment]
Probably push it this night. I'm not sure that it works perfect :(

**Comment by michalpolko** [comment]
I have mennaged to make Necrotic Plague "work" by hacking and I get a visual bug of displaying aura also as a buff which doesnt wear off ;/ But since I have no other solution I will use this, maybe somebody will come up with something better, maybe carlos93 will have something better :)

**Comment by carlos93** [comment]
I don't think so hahaha, but I wish to see your necrotic plague :D

**Comment by michalpolko** [comment]
Anybody knows:

1) What should be the damage of Soul Rip - channeled spell cast by Spirit Warden.
2) What should be the melee damage of Terenas and Spirit Warden.
3) What happens when Terenas dies? I know that Lich King gets the buff but does the player also die or Spirit Warden just starts attacking the player and player can still kill him before 60sec debuff wears off and kills the player?

**Comment by carlos93** [comment]
I think that the player have to die, none can "rescue" him from frostmourne. It's only a supposition ;)

**Comment by ghost** [comment]
my old guild videos of the lich king fight 

http://www.youtube.com/watch?v=6s3GwMBk6Bg part 1 
http://www.youtube.com/watch?v=Ry0UyqRUCFU part 2  ( this vid starts as soon as ported into frost mourne )
all texts and ect visble and audioable

might offer some help

**Comment by carlos93** [comment]
A problem that I found are if spell jumps and the target has necrotic plague, instead of adding 1 stack, it refresh the debuff and restarts the effect (50k damage, 0 stacks).

It's the same problem who has the warrior ability "sunder armor" that if one warrior casts 5 stacks (-20% armor) and other warrior cast sunder armor, don't refresh the 5 charges, and restarts only the debuff.

**Comment by michalpolko** [comment]
Thx for the video link, it will be useful.

About Necrotic Plague, there should be probably handled a few specific situations like if plague jumps on somebody who already has the aura or has the other aura (there are 2, one used by LK and the other one for "jumping"). i'll try to fix all of them.

**Comment by michalpolko** [comment]
https://github.com/przemratajczak/mangos/commit/593277dddd952cd1b1ccac5d3874eca3180b2c93
and
https://github.com/przemratajczak/scriptdev2/commit/a21d17c2248b192dd70434f42afe29df8bed82bf

should make Necrotic Plague stack after jumping from 2 different caster to 1 target and Plague shouldn't be casted on somebody that already has the plague.

**Comment by michalpolko** [comment]
What are the coords to Frostmourne Room?

**Comment by ghost** [comment]
Mich in the dbc s I remeber seening a table with safe locs and ect I bet u can find it there

**Comment by ghost** [comment]
http://getmangos.com/wiki/AreaTable.dbc. here is area id and ect I know I seen the safe locs for this area ill look more when I get home at work right now

**Comment by rsa** [comment]
i make two spells to teleport in FR. see on his...

**Comment by michalpolko** [comment]
I don't have any of the spells added into spell_target_position.

I found the place where should be frostmourne area (to the west from Frozen Throne) but there is no gameobject (platform) there. Or maybe I can't see it? What is the entry for it?

**Comment by rsa** [comment]
1.don't forget set visibility.distance.instances = 400 (or 500)
2. see in mr01212_mangos_area_spells.sql

**Comment by michalpolko** [comment]
OK, it works. I was missing config and the sql ;P thx

**Comment by Undergarun** [comment]
Checkout my feedback: https://github.com/mangosR2/scriptdev2/issues/76

**Comment by Undergarun** [comment]
Thank you for work.Please, bug reports and feedback about ICC in new issue.

**Comment by rsa** [comment]
need continue discussion in feedback issues.

**Comment by michalpolko** [comment]
Can somebody explain how switching between normal and heroic modes worked on late 3.3.5a in ICC? How ID on bosses and attempts were handled? Need info from some blizz servers player.

**Comment by michalpolko** [comment]
After some commit in core the spell of Sindragosa: Frost Bomb is now not working properly. Missile is launched and after a few seconds after missile hits the spot, the aura that marks the target destination appears and then after another few seconds the explosion occurs.
Any idea why it happens? I happens on old revision mr1505, on latest I dont have even the aura that marks target destination nor the damaging explosion, only missile visual ;/

**Comment by rsa** [comment]
may be removed my very old script on this spell dummy target?

**Comment by Venom** [comment]
missing ICECROWN CITADEL  +30% all stat BUFF.  

WORKING only icecrown citadel  -20% Dodge buff

i already applyed all scriptdev sql files  but still missing, players dont get that buff.

**Comment by Undergarun** [comment]
All ICC buffs in SQL commented, uncomment what percentage buff u want and execute.

**Comment by Venom** [comment]
sorry i dont understand what did u mean? by uncomment?

**Comment by Undergarun** [comment]
http://en.wiktionary.org/wiki/uncomment

scriptdev2 / sql_mr / mr00363_mangos_icecrown.sql

-- Icecrown buff
DELETE FROM `spell_area` WHERE `spell` IN
(73762, 73824, 73825, 73826, 73827, 73828, -- Alliance
73816, 73818, 73819, 73820, 73821, 73822);
-- +5%
-- INSERT INTO `spell_area` (`spell`, `area`, `racemask`, `gender`, `autocast`) VALUES
-- (73762, 4812, 1101, 2, 1), -- Alliance
-- (73816, 4812, 690, 2, 1); -- Horde
-- +10%
-- INSERT INTO `spell_area` (`spell`, `area`, `racemask`, `gender`, `autocast`) VALUES
-- (73824, 4812, 1101, 2, 1), -- Alliance
-- (73818, 4812, 690, 2, 1); -- Horde
-- +15%
-- INSERT INTO `spell_area` (`spell`, `area`, `racemask`, `gender`, `autocast`) VALUES
-- (73825, 4812, 1101, 2, 1), -- Alliance
-- (73819, 4812, 690, 2, 1); -- Horde
-- +20%
-- INSERT INTO `spell_area` (`spell`, `area`, `racemask`, `gender`, `autocast`) VALUES
-- (73826, 4812, 1101, 2, 1), -- Alliance
-- (73820, 4812, 690, 2, 1); -- Horde
-- +25%
-- INSERT INTO `spell_area` (`spell`, `area`, `racemask`, `gender`, `autocast`) VALUES
-- (73827, 4812, 1101, 2, 1), -- Alliance
-- (73821, 4812, 690, 2, 1); -- Horde
-- +30%
-- INSERT INTO `spell_area` (`spell`, `area`, `racemask`, `gender`, `autocast`) VALUES
-- (73828, 4812, 1101, 2, 1), -- Alliance
-- (73822, 4812, 690, 2, 1); -- Horde

**Comment by rsa** [comment]
blizz manually change this buff some times while 2010 year.

**Comment by Venom** [comment]
i'm interesting in what condition do u get this buffs?
because this buffs not working. i just updated server and they are not applyed to players...

**Comment by michalpolko** [comment]
After tracking last bug reports about ICC I must say that already a lot has changed in the repository and there were no adjustments for ICC related code after those changes so the instance is now probably working very badly ;/

Since I dont work with mangos anymore, I would like to ask if somebody is fixing anything related to ICC on their own fork and doesnt put those fixes into r2 repo, pls consider pushing/pull requesting those fixes, lets not let ICC rot slowly like it is right now.

**Comment by carlos93** [comment]
Im fixing details (timers,..) on my repo. I would to push it (with achievements ;)).
The problem its spells bugged...

**Comment by dufernst** [comment]
Are hack fixes considered? We currently have some of those but it's actually the spell system that needs to be redone.

**Comment by ZuSe** [comment]
michalpolko.. why dont u work with mangos anymore?
U did such a great job.. noone knows the icc-scripts better than you. Why dont give it a a try again?
As u said.. its sad to see icc going to die. I am a beginner with mangos-core so its hard for me to do the necessary changes, but i could provide you full DB support and did the EventAI-stuff for the trashmobs.

**Comment by michalpolko** [comment]
In last few weeks I continued my work on ICC and  fixed some bugs.
Currently known bugs are: not working AoE Mind Control auras in Deathwhisper and Lana'thel encounters and Putricide's Abomination can't be healed and in heroic mode (10 man at least) seems not to be summoned at all.

Would be nice to have feedback on how bosses are working now, results of some tests on live servers in 10 and 25man versions both would be nice to see ;)

**Comment by raynar** [comment]
Yes, we are going to test your ICC in 10 and 25 normal. But last core isn't stable, so we have to wait a little to update :(

**Comment by rsa** [comment]
after 12026 really, some unstability detected. need more tests/crushdumps.

**Comment by rsa** [comment]
first line in issue -

> It is no real issue, but since we have no r2 community forums... ;D

now R2 has community forum. really, many thing required fixing. i'm has one message (~20kb on russian), in where ICC compared with off step-by-step by expirienced player.
also need continue work under gunship battle. we may start this with static ship positions (movement be added later)?

**Comment by Undergarun** [comment]
I don't have nothing to commit but you have some feedback on issues and i'll provide more feedback if you want to work on finish ICC and let it as much playable as possible.

**Comment by rsa** [comment]
npc-on-transport currently not implemented (but i'm has at least 3 different worked realization :) if need, may be finished faster.

**Comment by Dokman** [comment]
@ michalpolko you have reason, if rsa don't implemented npc on transport is very impossible to make any kind of gunship

**Comment by Undergarun** [comment]
@michalpolko You still planning to continue working on ICC?

**Comment by Dokman** [comment]
@Undergarun can you see your personal mangos in issues ?

**Comment by michalpolko** [comment]
@Undergarun
in sd2 - yes
in mangosR2 - yes, if I finally manage to set up and run mangosR2 core on my local machine

I've been busy lately, if I was coding at home then it was only for work (can't keep on coding 24/7 ;p). Right now I'm finishing most of my projects and I will finally have some free time to spend on mangos.

**Comment by Dokman** [comment]
@michalpolko Good Job if you finish the ICC, i adore you :3
